File: app/src/main/java/com/ed/redditapp/ui/main/MainPostsAsyncTask.java
package com.ed.redditapp.ui.main;

import com.ed.redditapp.lib.AsyncTask;
import com.ed.redditapp.lib.api.Post;
import com.ed.redditapp.lib.api.RedditApi;

public class MainPostsAsyncTask {
    public interface Callback {
        void callback(Post[] posts);
    }

    private final RedditApi redditApi;

    public MainPostsAsyncTask(RedditApi redditApi) {
        this.redditApi = redditApi;
    }

    public void run(Callback callback) {
        new AsyncTask().runAsync((h) -> {
            Post[] posts = redditApi.getSubredditPosts("popular");
            h.onUIThread(() -> callback.callback(posts));
        });
    }
}

File: app/src/main/java/com/ed/redditapp/ui/main/MainPostsAsyncTask.kt
package com.ed.redditapp.ui.main

import com.ed.redditapp.lib.AsyncTask
import com.ed.redditapp.lib.api.Post
import com.ed.redditapp.lib.api.RedditApi

typealias MainPostsCallback = (posts: Array<Post>) -> Unit

class MainPostsAsyncTask(val redditApi: RedditApi) {
    fun run(after: String?, callback: MainPostsCallback) {
        AsyncTask().runAsync { h ->
            val posts = redditApi.getSubredditPosts("popular", after)
            h.onUIThread { callback(posts) }
        }
    }
}